### PR TITLE
Document and restructure libnetplan's public API symbols

### PR DESCRIPTION
**Pull request by slyon**
## Description
Please review commit-by-commit.

It's mostly adding Doxygen comments to our public functions.
But also dropping one public symbol from the API (`netplan_generate`), which is not properly implemented and still unused, so let's get rid of it until we grow a need for it and can have it implemented the correct way.

Finally, I'm moving around the documented declarations into their related header files:
- parser.h / parser-nm.h: Functions related to `NetplanParser*`
- state.h: Function related to `NetplanState*`
- netdef.h: Functions related to `NetplanNetDefinition*`
- utils.h: Function not related to any specific object
- types.h: data structures, no functions

#### Note
You can ignore the final commit ([abi-compat: Update for dropped 'netplan_generate' symbol](https://github.com/canonical/netplan/pull/438/commits/4bff7e6d481d0e1c73c717710abbbf5b66c1bff2)), which is auto-generated noise to `abi-compat/jammy_1.0.xml`.

FR-6079

## Checklist

- [x] Runs `make check` successfully.
- [x] Retains 100% code coverage (`make check-coverage`).
- [x] New/changed keys in YAML format are documented.
- [x] \(Optional\) Adds example YAML for new feature.
- [ ] \(Optional\) Closes an open bug in Launchpad.

### CODE REVIEW

File: include/netplan.h [hunk 1]
```diff
@@ -79,63 +151,185 @@ netplan_state_update_yaml_hierarchy(
...
+ * @details In the absence of a `match` stanza, the Netplan ID
+            can be considered to be the interface name of a single interface. Otherwise, it could match multiple interfaces.
+ * @param[in] netdef The @ref NetplanNetDefinition to query
+ * @return           Indication if @p netdef uses custom interface matching rules
+ */
 NETPLAN_PUBLIC gboolean
 netplan_netdef_has_match(const NetplanNetDefinition* netdef);
 
+/**
+ * @brief   Gheck if a @ref NetplanNetDefinition matches on given interface parameters.
```
Comment (daniloegea) [inline include/netplan.h]: Typo, s/Gheck/Check/

File: include/types.h [hunk 2]
```diff
@@ -62,14 +70,19 @@ typedef enum {
...
-typedef struct netplan_net_definition NetplanNetDefinition;
+/// Private data structure to contain validated Netplan configuration, ready for writing to disk.
+/// See @ref netplan_state_new and related accessor functions.
 typedef struct netplan_state NetplanState;
 
+/// Private data structure to contain individual settings per Netplan ID.
+/// See @ref netplan_state_get_netdef, @ref netplan_netdef_get_id and related accessor functions.
+typedef struct netplan_net_definition NetplanNetDefinition;
+
+/// Renderer backends, supported by Netplan.
```
Comment (daniloegea) [inline include/types.h:85]: Remove the comma?

File: include/types.h [hunk 3]
```diff
@@ -93,56 +115,51 @@ struct _NetplanStateIterator {
...
 };
 
-/*
- * Errors for domain NETPLAN_VALIDATION_ERROR
- *
- * VALIDATION_ERRORS are expected to contain only the YAML file name
- * where the error was found.
+/**
+ * @brief   Errors for domain @ref NETPLAN_VALIDATION_ERROR.
+ * @details Such are expected to contain only the YAML file name
```
Comment (daniloegea) [inline include/types.h:140]: "Such errors"?

File: include/util.h [hunk 4]
```diff
@@ -15,11 +16,10 @@
...
 
-/*! \file util.h
- *  \brief A set of high-level helper functions that can be used when working
- *  with datastructures of the library.
- *
- *  For example, it can be used to initialize iterators or clear error structures.
+/**
+ * @file util.h
+ * @brief A set of generic helper functions that can be used when working with
+ *        Netplan data structures. Like handling of @ref NetplanError data.
```
Comment (daniloegea) [inline include/util.h:22]: s/Like/Such as/?

**Review by daniloegea — APPROVED** [review]
Looks good. I just left a few suggestions and notes about typos.

**Comment by slyon** [comment]
Thanks, those were some nice finds! I fixed it all as suggested and rebased. Let's get this merged, once CI is green.

@rkratky Feel free to still leave your comments in here. If anything, I can fix it in a follow-up PR.

**Comment by slyon** [comment]
The NetworkManager CI is failing again... But it is unrelated to these changes and needs fixing independently in the CI images.

**Pull request merged**